packages/core/src/actions/contracts/readContracts.ts

```diff
@@ -63,18 +67,31 @@ export async function readContracts<Data extends any[] = Result[]>({
         })
         .flat() as Data
     }
-    return (await Promise.all(promises)).flat() as Data
+    return (await Promise.all(promises())).flat() as Data
   } catch (err) {
+    if (err instanceof ContractResultDecodeError) throw err
     if (err instanceof ContractMethodNoResultError) throw err
+    if (err instanceof ContractRevertedError) throw err
 
-    const promises = contracts.map((contract) =>
-      readContract({ ...contract, overrides }),
-    )
+    const promises = () =>
+      contracts.map((contract) => readContract({ ...contract, overrides }))
     if (allowFailure) {
-      return (await Promise.allSettled(promises)).map((result) =>
-        result.status === 'fulfilled' ? result.value : null,
-      ) as Data
+      return (await Promise.allSettled(promises())).map((result, i) => {
+        if (result.status === 'fulfilled') return result.value
+        const { addressOrName, functionName, chainId, args } = contracts[
+          i
+        ] as ReadContractsContract
+        const error = new ContractRevertedError({
+          addressOrName,
+          functionName,
+          chainId: chainId ?? chain.mainnet.id,
```

Comment:
Let's import `mainnet` here instead of getting it off of `chain` since `chain` pulls in a lot more than necessary.